Modules/socketmodule.c

```diff
@@ -3405,6 +3405,7 @@ sock_connect_impl(PySocketSockObject *s, void* Py_UNUSED(data))
     return 1;
 }
 
+/* Returns 0 on success, -1 with exception set on failure, or error code if not. */
```

Comment:
Let's expand that comment to make things clear. IMO, it's best to treat this as two separate error-handling schemes:
```suggestion
/* Common functionality for socket.connect and socket.connect_ex.
 *
 * If *raise* is set:
 * - On success, return 0.
 * - On any failure, return -1 with an exception set.
 * If *raise* is zero:
 * - On success, return 0.
 * - On connect() failure, return errno (without an exception set)
 * - On other error, return -1 with an exception set.
 *
 *   Note that -1 is a valid errno value on some systems.
 */
```